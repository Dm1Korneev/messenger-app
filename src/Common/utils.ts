
type Filter = Parameters<typeof Array.prototype.filter>[0]
export const distinct: Filter = (value, index, self) => self.indexOf(value) === index;
