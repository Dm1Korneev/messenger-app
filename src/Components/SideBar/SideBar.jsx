import React, { useEffect } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import AddBox from '@material-ui/icons/AddBox';

import { DRAWER_WIDTH, RELOAD_PERIOD } from 'Constants';
import Chat from 'Components/Chat';

const useStyles = makeStyles((theme) => ({
  drawer: {
    overflowX: 'hidden',
    whiteSpace: 'nowrap',
    position: 'relative',
  },
  drawerOpen: {
    width: DRAWER_WIDTH,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerClose: {
    width: theme.spacing(9),
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  toolbarIcon: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: theme.spacing(0, 1),
    ...theme.mixins.toolbar,
  },
  addChatIcon: {
    fontSize: theme.spacing(6),
    marginLeft: -theme.spacing(0.5),
  },
}));

const SideBar = ({
  getChats, drawerIsOpen,
  chats,
  activeChat,
  onDrawerClose,
  changeActiveChat,
  openAddChatDialog,
  openModifyChatDialog,
}) => {
  useEffect(() => {
    const interval = setInterval(getChats, RELOAD_PERIOD);
    return () => {
      if (interval) {
        clearInterval(interval);
      }
    };
  }, [getChats]);
  const classes = useStyles();

  return (
    <Drawer
      variant="permanent"
      classes={{
        paper: clsx(
          classes.drawer, {
            [classes.drawerOpen]: drawerIsOpen,
            [classes.drawerClose]: !drawerIsOpen,
          },
        ),
      }}
      open={drawerIsOpen}
    >
      <div className={classes.toolbarIcon}>
        <IconButton onClick={onDrawerClose}>
          <ChevronLeftIcon />
        </IconButton>
      </div>
      <Divider />
      <List>
        <ListItem button key="add_chat" onClick={openAddChatDialog}>
          <ListItemIcon>
            <AddBox color="primary" className={classes.addChatIcon} />
          </ListItemIcon>
          <ListItemText primary="Add chat" />
        </ListItem>
        {chats.map((chat) => {
          const { _id } = chat;
          return (
            <Chat
              chat={chat}
              key={_id}
              selected={_id === activeChat}
              chatOnClick={() => changeActiveChat({ activeChat: _id })}
              chatModifyOnClick={openModifyChatDialog}
            />
          );
        })}
      </List>
    </Drawer>
  );
};

SideBar.defaultProps = {
  activeChat: null,
  chats: [],
};
SideBar.propTypes = {
  getChats: PropTypes.func.isRequired,
  drawerIsOpen: PropTypes.bool.isRequired,
  chats: PropTypes.arrayOf(PropTypes.instanceOf(Object)),
  activeChat: PropTypes.string,
  onDrawerClose: PropTypes.func.isRequired,
  changeActiveChat: PropTypes.func.isRequired,
  openAddChatDialog: PropTypes.func.isRequired,
  openModifyChatDialog: PropTypes.func.isRequired,
};

export default SideBar;

