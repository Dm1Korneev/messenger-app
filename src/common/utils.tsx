export function distinct(value: any, index: number, self: Array<any>): boolean {
  return self.indexOf(value) === index;
}
