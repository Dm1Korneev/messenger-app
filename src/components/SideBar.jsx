import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import withStyles from '@material-ui/core/styles/withStyles';
import IconButton from '@material-ui/core/IconButton';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import Divider from '@material-ui/core/Divider';
import AddBoxIcon from '@material-ui/icons/AddBox';

import { DRAWER_WIDTH, RELOAD_PERIOD } from 'Common/constants';

import Chat from './Chat';

class SideBar extends React.Component {
  constructor(props) {
    super(props);

    this.getChats = props.getChats;
    this.interval = undefined;
  }

  componentDidMount() {
    this.interval = setInterval(this.getChats, RELOAD_PERIOD);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render() {
    const {
      drawerIsOpen,
      chats,
      activeChat,
      onDrawerClose,
      classes,
      changeActiveChat,
      openAddChatDialog,
      openModifyChatDialog,
    } = this.props;

    return (
      <Drawer
        variant="permanent"
        classes={{
          paper: classNames(
            classes.drawerPaper,
            !drawerIsOpen && classes.drawerPaperClose,
          ),
        }}
        open={drawerIsOpen}
      >
        <div className={classes.toolbarIcon}>
          <IconButton onClick={onDrawerClose}>
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <Divider />
        <List>
          <ListItem button key="add_chat" onClick={openAddChatDialog}>
            <AddBoxIcon color="primary" className={classes.addChatIcon} />
            <ListItemText primary="Add chat" />
          </ListItem>

          {chats.allIds.map((value) => {
            const chat = chats.byId[value];
            const { _id } = chat;
            return (
              <Chat
                chat={chat}
                key={_id}
                selected={_id === activeChat}
                chatOnClick={changeActiveChat}
                chatModifyOnClick={openModifyChatDialog}
              />
            );
          })}
        </List>
      </Drawer>
    );
  }
}

SideBar.defaultProps = {
  activeChat: null,
  chats: [],
};
SideBar.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  getChats: PropTypes.func.isRequired,
  drawerIsOpen: PropTypes.bool.isRequired,
  chats: PropTypes.shape({
    allIds: PropTypes.arrayOf(PropTypes.string),
    byId: PropTypes.object,
  }),
  activeChat: PropTypes.string,
  onDrawerClose: PropTypes.func.isRequired,
  changeActiveChat: PropTypes.func.isRequired,
  openAddChatDialog: PropTypes.func.isRequired,
  openModifyChatDialog: PropTypes.func.isRequired,
};

const styles = (theme) => ({
  drawerPaper: {
    position: 'relative',
    whiteSpace: 'nowrap',
    width: DRAWER_WIDTH,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerPaperClose: {
    overflowX: 'hidden',
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    width: theme.spacing.unit * 7,
    [theme.breakpoints.up('sm')]: {
      width: theme.spacing.unit * 9,
    },
  },
  toolbarIcon: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '0 8px',
    ...theme.mixins.toolbar,
  },
  addChatIcon: {
    fontSize: theme.spacing.unit * 7,
    marginLeft: -theme.spacing.unit,
  },
});

export default withStyles(styles)(SideBar);
