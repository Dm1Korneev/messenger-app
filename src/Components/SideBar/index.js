import commonHoc from 'Components/commonHoc';

import {
  changeActiveChat,
  getChats,
  openModifyChatDialog,
  setAddChatDialogIsOpen,
  setDrawerIsOpen,
} from 'Redux/actions';

import { chatsArraySelector } from 'Selectors/chats';
import { activeChatIdSelector, drawerIsOpenSelector } from 'Selectors/session';

import SideBar from './SideBar';

const mapStateToProps = (state) => ({
  chats: chatsArraySelector(state),
  activeChat: activeChatIdSelector(state),
  drawerIsOpen: drawerIsOpenSelector(state),
});

const mapDispatchToProps = {
  onDrawerClose: () => setDrawerIsOpen(false),
  changeActiveChat,
  openAddChatDialog: () => setAddChatDialogIsOpen(true),
  openModifyChatDialog,
  getChats,
};

const Container = commonHoc(SideBar, {
  mapStateToProps,
  mapDispatchToProps,
});

export { Container as SideBar };
