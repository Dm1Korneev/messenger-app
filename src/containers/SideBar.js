import { connect } from "react-redux";
import {
  setDrawerIsOpen,
  changeActiveChat,
  setAddChatDialogIsOpen,
  openModifyChatDialog,
  reloadChatsList
} from "../redux/actions";
import SideBar from "../components/SideBar";

const mapStateToProps = state => ({
  chats: state.chats,
  activeChat: state.session.activeChat,
  drawerIsOpen: state.session.drawerIsOpen
});

const mapDispatchToProps = dispatch => ({
  onDrawerClose: () => dispatch(setDrawerIsOpen(false)),
  changeActiveChat: activeChat => dispatch(changeActiveChat(activeChat)),
  openAddChatDialog: () => dispatch(setAddChatDialogIsOpen(true)),
  openModifyChatDialog: chat => dispatch(openModifyChatDialog(chat)),
  reloadChatsList: () => dispatch(reloadChatsList())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SideBar);
