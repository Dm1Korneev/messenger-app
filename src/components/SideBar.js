import React, { Component } from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import IconButton from "@material-ui/core/IconButton";
import classNames from "classnames";
import Drawer from "@material-ui/core/Drawer";
import ListItem from "@material-ui/core/ListItem";
import List from "@material-ui/core/List";
import ListItemText from "@material-ui/core/ListItemText";
import ChevronLeftIcon from "@material-ui/icons/ChevronLeft";
import Divider from "@material-ui/core/Divider";
import { DRAWER_WIDTH } from "../common/constants";
import AddBoxIcon from "@material-ui/icons/AddBox";
import Chat from "./Chat";

import { RELOAD_PERIOD } from "../common/constants";

class SideBar extends Component {
  constructor(props) {
    super(props);

    this.reloadChatsList = props.reloadChatsList;
    this.interval = undefined;
  }

  componentDidMount() {
    this.interval = setInterval(this.reloadChatsList, RELOAD_PERIOD);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render() {
    const {
      drawerIsOpen,
      chats,
      activeChat,
      onDrawerClose,
      classes,
      changeActiveChat,
      openAddChatDialog,
      openModifyChatDialog
    } = this.props;

    return (
      <Drawer
        variant="permanent"
        classes={{
          paper: classNames(
            classes.drawerPaper,
            !drawerIsOpen && classes.drawerPaperClose
          )
        }}
        open={drawerIsOpen}
      >
        <div className={classes.toolbarIcon}>
          <IconButton onClick={onDrawerClose}>
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <Divider />
        <List>
          <ListItem button key={"add_chat"} onClick={openAddChatDialog}>
            <AddBoxIcon color="primary" className={classes.addChatIcon} />
            <ListItemText primary={"Add chat"} />
          </ListItem>

          {chats.allIds.map(value => {
            const chat = chats.byId[value];
            const { _id } = chat;
            return (
              <Chat
                chat={chat}
                key={_id}
                selected={_id === activeChat}
                chatOnClick={changeActiveChat}
                chatModifyOnClick={openModifyChatDialog}
              />
            );
          })}
        </List>
      </Drawer>
    );
  }
}

const styles = theme => ({
  drawerPaper: {
    position: "relative",
    whiteSpace: "nowrap",
    width: DRAWER_WIDTH,
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  drawerPaperClose: {
    overflowX: "hidden",
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    }),
    width: theme.spacing.unit * 7,
    [theme.breakpoints.up("sm")]: {
      width: theme.spacing.unit * 9
    }
  },
  toolbarIcon: {
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-end",
    padding: "0 8px",
    ...theme.mixins.toolbar
  },
  addChatIcon: {
    fontSize: theme.spacing.unit * 7,
    marginLeft: -theme.spacing.unit
  }
});

export default withStyles(styles)(SideBar);
