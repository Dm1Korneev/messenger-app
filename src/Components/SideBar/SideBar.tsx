import AddBox from '@mui/icons-material/AddBox';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import Box from '@mui/material/Box';
import Divider from '@mui/material/Divider';
import Drawer from '@mui/material/Drawer';
import IconButton from '@mui/material/IconButton';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';

import { Chat } from 'Components/Chat';
import { DRAWER_WIDTH } from 'Constants';
import { useChats } from 'Hooks';

type SideBarProps = {
  chatModifyOnClick: (id: string) => void
  onChatAddClick: () => void
  isDrawerOpen: boolean
  onDrawerClose: () => void
  activeChatId?: string
  onChatClick: (chatId: string) => void
}

export const SideBar = ({
  chatModifyOnClick, onChatAddClick, isDrawerOpen, onDrawerClose, activeChatId, onChatClick,
}: SideBarProps) => {
  const { data: chats } = useChats();

  return (
    <Drawer
      open={isDrawerOpen}
      sx={(theme) => {
        const additionalStyles = isDrawerOpen ? {
          width: DRAWER_WIDTH,
          transition: theme.transitions.create('width', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.enteringScreen,
          }),
        } : {
          width: theme.spacing(9),
          transition: theme.transitions.create('width', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen,
          }),
        };

        return {
          '.MuiDrawer-paper': {
            overflowX: 'hidden',
            whiteSpace: 'nowrap',
            position: 'relative',
            ...additionalStyles,
          },
        };
      }}
      variant="permanent"
    >
      <Box sx={(theme) => ({
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'flex-end',
        p: theme.spacing(0, 1),
        ...theme.mixins.toolbar,
      })}
      >
        <IconButton onClick={onDrawerClose} size="large">
          <ChevronLeftIcon />
        </IconButton>
      </Box>
      <Divider />
      <List>
        <ListItem key="add_chat" button onClick={onChatAddClick}>
          <ListItemIcon>
            <AddBox
              color="primary"
              sx={(theme) => ({
                fontSize: theme.spacing(6),
                ml: -0.5,
              })}
            />
          </ListItemIcon>
          <ListItemText primary="Add chat" />
        </ListItem>
        {chats.map((chat) => {
          const { _id } = chat;
          return (
            <Chat
              key={_id}
              chat={chat}
              chatModifyOnClick={chatModifyOnClick}
              chatOnClick={() => onChatClick(_id)}
              selected={_id === activeChatId}
            />
          );
        })}
      </List>
    </Drawer>
  );
};
