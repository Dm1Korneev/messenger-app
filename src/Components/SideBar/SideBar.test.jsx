import React from 'react';
import { render } from '@testing-library/react';

import { mockComponent } from 'testing/utils';

import SideBar from './SideBar';

jest.mock('Components/Chat', () => mockComponent('Chat'));
jest.mock('Components/UserNameField', () => mockComponent('UserNameField'));
jest.mock('Components/UserPasswordField', () => mockComponent('UserPasswordField'));
jest.mock('Components/AvatarSelector', () => mockComponent('AvatarSelector'));

jest.mock('@material-ui/core/IconButton', () => mockComponent('IconButton'));
jest.mock('@material-ui/core/Drawer', () => mockComponent('Drawer'));
jest.mock('@material-ui/core/ListItem', () => mockComponent('ListItem'));
jest.mock('@material-ui/core/List', () => mockComponent('List'));
jest.mock('@material-ui/core/ListItemText', () => mockComponent('ListItemText'));
jest.mock('@material-ui/core/Divider', () => mockComponent('Divider'));

jest.mock('@material-ui/icons/ChevronLeft', () => mockComponent('ChevronLeft'));
jest.mock('@material-ui/icons/AddBox', () => mockComponent('AddBox'));

const props = {
  getChats: jest.fn(),
  drawerIsOpen: false,
  onDrawerClose: jest.fn(),
  changeActiveChat: jest.fn(),
  openAddChatDialog: jest.fn(),
  openModifyChatDialog: jest.fn(),
  onSignIn: jest.fn(),
  onRegister: jest.fn(),
};

describe('render SideBar component', () => {
  const setup = () => render(
    <SideBar {...props} />,
  );

  test('snapshot test', async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});

