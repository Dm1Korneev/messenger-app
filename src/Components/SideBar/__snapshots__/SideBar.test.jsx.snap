// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SideBar component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="Drawer"
    test-prop-classes="{\\"paper\\":\\"SideBar-drawerPaper-1 SideBar-drawerPaperClose-2\\"}"
    test-prop-open="false"
    test-prop-variant="\\"permanent\\""
  >
    <div
      class="SideBar-toolbarIcon-3"
    >
      <div
        data-originalcomponent="IconButton"
      >
        <div
          data-originalcomponent="ChevronLeft"
        />
      </div>
    </div>
    <div
      data-originalcomponent="Divider"
    />
    <div
      data-originalcomponent="List"
    >
      <div
        data-originalcomponent="ListItem"
        test-prop-button="true"
      >
        <div
          data-originalcomponent="AddBox"
          test-prop-classname="\\"SideBar-addChatIcon-4\\""
          test-prop-color="\\"primary\\""
        />
        <div
          data-originalcomponent="ListItemText"
          test-prop-primary="\\"Add chat\\""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
